src/core/metrics.cc

```diff
@@ -163,205 +182,270 @@ Metrics::EnableGPUMetrics()
-      LOG_WARNING << "failed to get device properties for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
-
-    char pcibusid_str[64];
-    cudaerr =
-        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get Bus ID for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
+  if (standalone_) {
+    dcgmerr = dcgmUpdateAllFields(dcgm_handle_, 1);
+    if (dcgmerr != DCGM_ST_OK) {
+      LOG_WARNING
+          << "error, DCGM unable to update all fields, GPU metrics will "
+             "not be available: "
+          << errorString(dcgmerr);
+      return false;
     }
+  }
 
-    nvmlDevice_t gpu;
-    nvmlReturn_t nvmlerr = nvmlDeviceGetHandleByPciBusId_v2(pcibusid_str, &gpu);
-    if (nvmlerr != NVML_SUCCESS) {
-      LOG_WARNING << "failed to get device from Bus ID "
-                  << ", GPU metrics will not be available for this device: "
-                  << nvmlErrorString(nvmlerr);
-      continue;
-    }
+  unsigned int all_gpu_ids[DCGM_MAX_NUM_DEVICES];
+  int gpu_count;
+  dcgmerr = dcgmGetAllDevices(dcgm_handle_, all_gpu_ids, &gpu_count);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "error, DCGM unable to get device info and count, GPU "
+                   "metrics will not be available: "
+                << errorString(dcgmerr);
+    return false;
+  }
 
-    char gpu_name[NVML_DEVICE_NAME_BUFFER_SIZE + 1];
-    nvmlerr = nvmlDeviceGetName(gpu, gpu_name, NVML_DEVICE_NAME_BUFFER_SIZE);
-    if (nvmlerr == NVML_SUCCESS) {
-      LOG_INFO << "Collecting metrics for GPU " << didx << ": " << gpu_name;
+  // Get CUDA-visible PCI bus ids
+  std::set<std::string> pciBusIds;
+  for (int i = 0; i < gpu_count; ++i) {
+    std::string pciBusId = "0000";  // pad 0's for uniformity
+    char pcibusid_str[64];
+    cudaError_t cudaerr =
+        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, i);
+    if (cudaerr == cudaSuccess) {
+      pciBusId.append(pcibusid_str);
+      pciBusIds.emplace(pciBusId);
     } else {
-      LOG_INFO << "Collecting metrics for GPU " << didx;
+      LOG_WARNING << "error, CUDA unable to get PCI bus information for GPU:"
+                  << i;
     }
+  }
 
-    std::string uuid;
-    char uuid_str[NVML_DEVICE_UUID_BUFFER_SIZE + 1];
-    nvmlerr = nvmlDeviceGetUUID(gpu, uuid_str, NVML_DEVICE_UUID_BUFFER_SIZE);
-    if (nvmlerr == NVML_SUCCESS) {
-      uuid = uuid_str;
+  // Get DCGM metrics for each GPU. Some devices may have problems using DCGM
+  // API and thus these devices needs to be ignored.
+  std::vector<uint32_t> available_gpu_ids;
+  dcgmDeviceAttributes_t gpu_attributes[DCGM_MAX_NUM_DEVICES];
+  for (int i = 0; i < gpu_count; i++) {
```

Comment:
You are iterating through the DCGM devices and reportin DCGM device IDs. But that is not what the user understands. They understand CUDA device IDs. So this code seems a little backwards. Think about creating a map from pcibusid to DCGM device ID. Then iterate through the cuda devices and for each look up the DCGM id and from that the DCGM info. If you can't find DCGM id for a pcibusid then you can report the CUDA device id as not having metrics available.

Reply:
Should we use the CUDA id for all devices then? Right now, we are reporting all device ids from the DCGM standpoint.

Reply:
Yes. The user does not known anything about dcgm numbering of the GPUs. If they understand anything it will be the CUDA number of the GPUs. For example, if they run on a system with multiple GPUs but only one is visible to Triton, then they expect that GPU to be referred to as GPU 0, no matter what the DCGM id of that GPU is.